[chore] add logic to update our tests k8s versions

dloucasfx opened the pull request:
**Description:** <Describe what has changed.>
Added a utility to get a list of supported vanilla k8s clusters, compare it to our test k8s versions and update when necessary.

I added a make target, which we can keep or remove, but the end goal is to add a gh action that uses this logic to create a PR with the changes if detected.

We currently have a bunch of EOL k8s versions that we use, it's a way to check compatibility with cloud providers that extend the support. 
This is not ideal as we keep increasing our test matrix and the vanilla cluster is not identical to the one used by the cloud provider. To test for schema compatibility, we should start using `kubeconform`

To Do (sperate PRs):
- Add a gh action
- track EKS, AKS and GKE

**Testing:** <Describe what testing was performed and which tests were added.>
unit test

jinja2 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,142 @@
+				break
+			} else if i == len(versionBlock)-1 && eolDate.After(now) {
+				// Add the new version at the end of the list
+				output = append(output, fmt.Sprintf("%s- v%s # EOL %s", strings.Repeat(" ", len(line)-len(trimmed)), version.Latest, version.EOLDate))
```
We are appending to the list of kind cluster versions we run functional tests against, right? If that's the case we should confirm that kind has an image out for the new version we are adding here or we'll have errors when we try to start in the kind cluster.

dloucasfx replied:
Good point @jinja2 . I will add some checks and if not available go down to previous minor version

dloucasfx replied:
PR updated; we now check docker hub if the image exists and decrement the minor minor until we find one

pjanotti commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,222 @@
+package main
```
nit - I'm not sure about this but other go sources in the repo have a header. That said there is no check enforcing that....

pjanotti replied:
```suggestion
// Copyright Splunk Inc.
// SPDX-License-Identifier: Apache-2.0

package main
```

dloucasfx replied:
makes sense!

pjanotti commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,222 @@
+	k8sVersions, err := fetchKubernetesVersions(url)
+	if err != nil {
+		fmt.Println("Failed to get k8s versions:", err)
+		return
```
It will be good to set the exit code in this case so make and other tools could be notified of the failure.

dloucasfx replied:
Makes sense, I will be using
0 // success
1 // failed
2 // partial success. ie some files got update

pjanotti commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,222 @@
+		return
+	}
+
+	filesToUpdate := []string{".github/workflows/functional_test.yaml", ".github/workflows/functional_test_v2.yaml"}
```
nit: nice to have these as parameters, but, it can be as it is too

dloucasfx replied:
I am planning on expanding this to include EKS, AKS and GKE with an option to choose what cloud provider list to update and I will also add the option to pass the list of yaml file to update and the name of the array element where the versions exist. 
But this will be in the next version

pjanotti commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,222 @@
+
+		if inK8sVersionBlock {
+			if strings.HasPrefix(trimmed, "-") || strings.HasPrefix(trimmed, "#") {
+				k8sVersionBlock = append(k8sVersionBlock, line)
```
nit: a comment describing that here the code is collecting all entries in the list of version or some to that effect

pjanotti reviewed (commented):
Nice!

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,238 @@
+	return output, nil
+}
+
+func getAvailableKindImage(tag string) (string, error) {
```
This logic is pretty hard for me to follow. It looks like for every k8s version we're getting an available kind image. Is there some way to iterate over the kind images supported, rather than manually doing string operations to deprecate versions and check if they're valid? If we [ordered results simply by name](https://hub.docker.com/v2/repositories/kindest/node/tags?=page=1&ordering=name), we'd be able to just iterate directly checking if the k8s version exists as a kind version, and if not, keep iterating. We could also use [go-version](https://pkg.go.dev/github.com/hashicorp/go-version) to handle some of the versioning string/regex functionality comparing values if it's useful.

dloucasfx replied:
This was also drastically simplified in the next iteration.
But the idea of getting the kind images first is not ideal, as you will also have to do with paging and large payloads as there are many images/tags for different platform etc...
The idea here and still in the next iteration, that we have a list of "currently" supported k8s versions, with latest and cycle (those are normally 3 or 4 max versions), we iterate through the latest and decrement it until we find an existing image on docker hub and use that tag.

crobert-1 reviewed (commented):
Looks good! With all of my comments feel free to drop or clarify as you see fit. I'm generally trying to reduce regexes and string parsing logic as much as possible. I'd be happy to discuss any options here.

dloucasfx commented on .github/workflows/update_k8s_versions.yaml:
```diff
@@ -0,0 +1,41 @@
+
+on:
+  pull_request:
+#schedule:
+    # Run every Monday at noon.
+  #  - cron: "0 12 * * 1"
+  #workflow_dispatch:
```
correct, I was just testing it on pull_request, I'll update it soon

dloucasfx commented on .github/workflows/update_k8s_versions.yaml:
```diff
@@ -0,0 +1,41 @@
+      - name: check for changes
+        id: git-check
+        run: |
+          if git diff --quiet; then
```
`--exit-code` will show the diff , which is not what we need

`--quiet` prevents the diff from showing and exits with 1 if there was a diff, 0 if not
`--exit-code` will exit with 1 if there’s a diff and 0 if not but it’ll show the diff

dloucasfx commented on .github/workflows/update_k8s_versions.yaml:
```diff
@@ -0,0 +1,41 @@
+
+      - name: Open PR for K8S Version Update
+        if: steps.git-check.outputs.changes == 'true'
+        uses: peter-evans/create-pull-request@v7
+        with:
+          commit-message: Update K8S test versions
+          title: Update K8S versions used for testing
+          body: Use latest supported k8s kind versions
+          branch: update-k8s-test-versions
+          base: main
+          delete-branch: true
```
I don't want to deal with tokens/secret and all the extra steps and logic that `create-pull-request` takes care of for you. In addition to the token , the PR branch management https://github.com/marketplace/actions/create-pull-request
We also widely use this action in other workflows.

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,238 @@
+		trimmed := strings.TrimSpace(line)
+
+		// Locate the k8s-version block needed to update
+		if strings.HasPrefix(trimmed, "k8s-version:") || strings.HasPrefix(trimmed, "kubernetes_version:") {
```
In the coming iteration, I had to move the the matrix out of the workflow.
I will update the PR shortly and please review again, would love to hear your feedback

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,238 @@
+	return nil
+}
+
+func updateVersionsBlock(versions []KubernetesVersion, versionBlock []string) ([]string, error) {
```
this was drastically reduce in the next iteration,  I will update the PR shortly and please review again, would love to hear your feedback

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,238 @@
+	for _, filePath := range filesToUpdate {
+		filePath = filepath.Join(currentDir, filepath.Clean(filePath))
+		if err := updateK8sVersion(filePath, k8sVersions); err != nil {
+			fmt.Println("Error:", err)
+			partialFailure = true
```
I will add this! Thanks!

dloucasfx replied:
not needed with new iteration

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+func main() {
+	url := "https://endoflife.date/api/kubernetes.json"
+	k8sVersions, err := fetchKubernetesVersions(url)
```
good point

dloucasfx replied:
On a second thought, I need to pass the URL for unit tests when using a mock server

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,238 @@
+
+func imageTagExists(tag string) (bool, error) {
+	url := "https://hub.docker.com/v2/repositories/kindest/node/tags?page_size=1&page=1&ordering=last_updated&name="
+	resp, err := http.Get(url + tag)
```
fixed in the next iteration

dloucasfx commented on .github/workflows/update_k8s_versions.yaml:
```diff
@@ -0,0 +1,41 @@
+      - name: check for changes
+        id: git-check
+        run: |
+          if git diff --quiet; then
+            echo "No changes detected, exiting workflow successfully"
+            exit 0
+          fi
+          echo "changes=true" >> $GITHUB_OUTPUT
```
I am not sure I follow, but I am also a beginner when it comes to gh actions

dloucasfx replied:
@crobert-1 
Thanks, I see now, you want me to use the outcome property. 
The reason I have not followed this approach, because exiting with non-zero will mark the step as failed, and of course, still need to add `continue-on-error: true`

example:
 ````
       - name: check for changes
        id: git-check
        continue-on-error: true
        run: |
          if git diff --quiet; then
            echo "No changes detected, exiting workflow successfully"
            exit 0
          fi
          exit 1

      - name: Open PR for matrix version update
        if: steps.git-check.outcome == 'failure'
        run: |
          echo "create PR logic"
````

The exit 1 is needed to differentiate the states (change vs no-change).  (or simply replace the run with `git diff --quiet` , but still exit 1 applies)

If someone not familiar with how those steps work in gh action, looking at the log, you would think that the step has failed and this can be confusing
````
[Check for new matrix versions and update tests if needed/check_and_update] ⭐ Run Main check for changes
[Check for new matrix versions and update tests if needed/check_and_update]   🐳  docker exec cmd=[bash -e /var/run/act/workflow/git-check] user= workdir=
[Check for new matrix versions and update tests if needed/check_and_update] Failed but continue next step
[Check for new matrix versions and update tests if needed/check_and_update]   ❌  Failure - Main check for changes [127.233375ms]
[Check for new matrix versions and update tests if needed/check_and_update] ⭐ Run Main Open PR for matrix version update
[Check for new matrix versions and update tests if needed/check_and_update]   🐳  docker exec cmd=[bash -e /var/run/act/workflow/3] user= workdir=
| create PR logic
[Check for new matrix versions and update tests if needed/check_and_update]   ✅  Success - Main Open PR for matrix version update [69.318834ms]
[Check for new matrix versions and update tests if needed/check_and_update] ⭐ Run Complete job
[Check for new matrix versions and update tests if needed/check_and_update] Cleaning up container for job check_and_update
[Check for new matrix versions and update tests if needed/check_and_update]   ✅  Success - Complete job
[Check for new matrix versions and update tests if needed/check_and_update] 🏁  Job succeeded
````

In short, `steps.git-check.outcome` can create confusion as the step will be marked as failed and the current logic avoid this behavior

crobert-1 replied:
Okay, thanks for clarifying! Agreed that's not what we want, and I appreciate the job output too 👍

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	var testMatrix map[string]map[string][]string
+	if err := json.Unmarshal(content, &testMatrix); err != nil {
```
Same as other comment `err` ends up getting overshadowed here, we should use `=` instead through the rest of the method.

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,285 @@
+
+	for _, value := range testMatrix {
+		if len(kindVersions) > 0 && value["k8s-kind-version"] != nil {
+			value["k8s-kind-version"] = kindVersions
```
I'd prefer to have the json keys as constants that are reused rather than plain strings, just in case of typos or future changes.

dloucasfx commented on .github/workflows/update_k8s_versions.yaml:
```diff
@@ -0,0 +1,47 @@
+  schedule:
+    # Run every Monday at noon.
+    - cron: "0 12 * * 1"
+  workflow_dispatch:
+    inputs:
+      DEBUG_ARGUMENT:
+        description: 'Enable debug by setting -debug to true'
+        required: false
+        default: '-debug=false'
```
Right, you can run the debug locally. But this is just to give the user an option to debug the gh action in case (I can't think of one now, maybe when extra logic is added later on), you can't reproduce locally.

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,285 @@
+	return body, nil
+}
+
+func logDebug(format string, v ...interface{}) {
```
Yeah, I was considering zap package , but for a small tool like this, the standard package log would the job .
If it's ok with you, I would like to stick with log.

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,285 @@
+
+	now := time.Now()
+	for _, kubernetesVersion := range kubernetesVersions {
+		eolDate, err := time.Parse("2006-01-02", kubernetesVersion.EOLDate)
```
sounds good, I can do that!

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	// Parse and cleanup the slice values
+	minikuveVersions := strings.Split(strings.ReplaceAll(strings.ReplaceAll(matches[1], "\n", ""), `"`, ""), ",")
```
yuck :) , good catch!

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,285 @@
+		return nil, fmt.Errorf("failed to get k8s versions: %w", err)
+	}
+	var kubernetesVersions, supportedKubernetesVersions []KubernetesVersion
+	if err := json.Unmarshal(body, &kubernetesVersions); err != nil {
```
good catch!

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	// Parse and cleanup the slice values
+	minikuveVersions := strings.Split(strings.ReplaceAll(strings.ReplaceAll(matches[1], "\n", ""), `"`, ""), ",")
```
I run `strings.TrimSpace(minikubeVersion)` when I build the returned slice.
But I guess I cam move it to this code as we print the output

dloucasfx replied:
Ended up using replacer , which is cleaner to read

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+
+	now := time.Now()
+	for _, kubernetesVersion := range kubernetesVersions {
+		eolDate, err := time.Parse(time.DateOnly, kubernetesVersion.EOLDate)
```
This is another shadowing error

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	var kindImage DockerImage
+	if err := json.Unmarshal(body, &kindImage); err != nil {
```
This is another shadowing error

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+		}
+	}
+	// Marshal the updated test matrix back to JSON
+	updatedContent, err := json.MarshalIndent(testMatrix, "", "  ")
```
This is a shadowing error

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
```
This is a shadowing error

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+	}
+	logDebug("Found supported k8s versions %v", k8sVersions)
+
+	kindVersions, err := getLatestSupportedKindImages(KindDockerHubURL, k8sVersions)
```
This is a shadowing error

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+		logDebug("Found supported kind images: %v", kindVersions)
+	}
+
+	minikubeVersions, err := getLatestSupportedMinikubeVersions(MiniKubeURL, k8sVersions)
```
This is a shadowing error

crobert-1 commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	path := "ci-matrix.json"
+	currentDir, err := os.Getwd()
```
This is a shadowing error

crobert-1 reviewed (commented):
Sorry about all of the nits, once the formatting tools are introduced to the repo it'll catch earlier on in the process.

dmitryax commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,287 @@
+		logDebug("Found supported minikube versions: %v", minikubeVersions)
+	}
+
+	if len(kindVersions) == 0 && len(minikubeVersions) == 0 {
```
Shouldn't we fail if any of them is empty not both?

dloucasfx replied:
I figured if one provider failed (ex: the endpoint was not reachable), we should not sacrifice the other

dloucasfx commented:
New push:
- Addressed most of the review entries. For var shadowing, I only addressed the linter errors as the others are not necessary
- Updated the golangci config to force strict shadow detection and excluded this tool from gosec to prevent unnecessary detections 
```
tools/k8s_versions/update_k8s_versions.go:197:11: G306: Expect WriteFile permissions to be 0600 or less (gosec)
        if err = os.WriteFile(filePath, updatedContent, 0644); err != nil {
                 ^
tools/k8s_versions/update_k8s_versions.go:218:15: G107: Potential HTTP request made with variable url (gosec)
        resp, err := http.Get(url)
      
```
-  Rebased and resolved conflicts

crobert-1 commented:
Once this PR has been rebased on `main` we can drop all of my `shadowing err` issues. `main` now has lint checks running on PRs so once lint passes on this PR my concerns are resolved 👍

dloucasfx commented:
> Once this PR has been rebased on `main` we can drop all of my `shadowing err` issues. `main` now has lint checks running on PRs so once lint passes on this PR my concerns are resolved 👍

@crobert-1 
Pushed the following changes:
- rebased
- Added Makefile in the directory to reference the common makefile
- Addressed @dmitryax comment https://github.com/signalfx/splunk-otel-collector-chart/pull/1792#discussion_r2072150118
- Fixed gosec G107 by using url parse
```
tools/k8s_versions/update_k8s_versions.go:220:15: G107: Potential HTTP request made with variable url (gosec)
        resp, err := http.Get(url)
```
- The only lint that is failing now is 
```
tools/k8s_versions/update_k8s_versions.go:199:11: G306: Expect WriteFile permissions to be 0640 or less (gosec)
        if err = os.WriteFile(filePath, updatedContent, 0o644); err != nil {
```
This is very restrictive, not if we want to no lint it or change the G306 gosec config to 
```
    config:
      # Maximum allowed permissions mode for os.OpenFile and os.Chmod
      # Default: "0600"
      G306: "0644"
```

crobert-1 commented:
> This is very restrictive, not if we want to no lint it or change the G306 gosec config to

I'd say for now my preference is to just add a nolint comment. I generally prefer getting warnings up front on new changes to let us make an educated decision on a case by case basis. Happy to hear from others though 👍

crobert-1 reviewed (approved):
LGTM once CI is passing. Thanks for all of your work here @dloucasfx!

dloucasfx commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,292 @@
+
+	logDebug("Found supported minikube versions: %v", minikubeVersions)
+
+	if len(kindVersions) == 0 || len(minikubeVersions) == 0 {
```
done

dmitryax commented on tools/k8s_versions/update_k8s_versions.go:
```diff
@@ -0,0 +1,294 @@
+	return false, nil
+}
+
+func decrementMinorMinorVersion(version string) (string, error) {
```
It's called the patch version in semver :) 
```suggestion
func decrementPatchVersion(version string) (string, error) {
```

dloucasfx replied:
roger that!

jinja2 commented:
@dloucasfx  We have some tests which use the K8S_VERSION var to look up for expected test data since the metrics change with k8s version for some receiver. These tests (i can only remember histograms from the top of my head) will probably start failing once a new minor version is introduced by this workflow. I took a look at the code and don't see this being addressed. Is this known? we can fix the tests once the new version is added or we could probably just copy the data from `new minor version - 1` when the workflow runs.
